### ISSUE TITLE
Expose readiness stream on `TcpStream`

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio#181]
The goal is to support the caller being able to detect a HUP event without requiring reading.

Currently, `PollEvented::poll_read_ready` returns the entire set of readiness bits. If any are present, the current task is not notified when the readiness set changes. Specifically, if the socket is currently readable, and a HUP event is received, the task is not notified. This prevents being able to reliably detect HUP events.

I see two ways to solve this. They are only subtly different.

### Option 1

The first, is to update `poll_read_ready` to only return `Ready(bits)` when a readiness event has been returned. So, if a socket is currently read ready:

```rust
io.poll_read_ready()? // => Ready(readable)
io.poll_read_ready()? // => NotReady
```

This would require adding a new fn that returns the current readiness state without registering to be notified:

```rust
fn get_read_ready(&self) -> mio::Ready { ... }
```

In order to detect HUP events, one would do the following:

```rust
while !try_ready!(io.poll_read_ready()).is_hup() {}
// Handle HUP
```

### Option 2

The other option is to have `poll_read_ready` *always* register the current task to be notified when a readiness event is received even if `NotReady` is **not** returned.

This would lead to a simpler API at the cost of potentially more task notifications. This also might not be an idiomatic API as the task is notified w/o `NotReady` returned. However, I don't think it is a violation of the contract.

This relates to https://github.com/runconduit/conduit/issues/511.

**Comment by olix0r** [comment]
👍 to the general plan of exposing readiness as a stream

## Option 1

IIUC, _Option 1_ cannot guarantee that all readiness states are observable. I.e. to epoll events can fire, such that the second notification hides the first. I defer to those with more familiarity of these APIs wrt whether this is an actual concern.

## Option 2

> This would lead to a simpler API at the cost of potentially more task notifications.

It doesn't seem like additional task notification is a critical concern, in practice... it's pretty rare that you only want to poll something once, leaving it in ready.

But this option is a bit surprising, so we may want to avoid it purely for ergonomics.

---

Additionally: does this need to be reflected in `AsyncRead`? Should there be a generic way to get notified of stream hangups?

**Comment by carllerche** [comment]
I think potentially losing some intermediate mio events is fine as long as the final return value represents all the readiness (ORing all the values together).

> Additionally: does this need to be reflected in `AsyncRead`? Should there be a generic way to get notified of stream hangups?

If it needs to be represented by a trait, it probably should be a standalone trait. HUP really means both read AND write closed and doesn't really fit in with either.

---

Thinking more about this, both option A would require more ops than necessary to impl `poll_hup` like functionality:

```rust
fn poll_hup(io: TcpStream) -> Poll<(), io::Error> {
    if io.get_read_ready().is_hup() {
        return Ok(().into());
    }

    while !try_ready!(io.poll_read_ready()).is_hup() {}

    Ok(().into())
}
```

The initial `get_read_ready` check is to catch if the HUP has already been observed by the socket.

I also have a third option:

### Option 3

```rust
fn poll_read_ready(&self, mask: Ready) -> Poll<Ready, io::Error> {
    // ...
}
```

This behaves similarly to `poll_read_ready` today. If a socket is currently read ready, then:

```rust
io.poll_read_ready(Ready::readable())? // => Ready(readable)
io.poll_read_ready(Ready::readable())? // => Ready(readable)
io.need_read();
io.poll_read_ready(Ready::readable())? // => NotReady, task registered
```

Then, the poll_hup fn becomes:

```rust
fn poll_hup(io: TcpStream) -> Poll<(), io::Error> {
    while !try_ready!(io.poll_read_ready(Ready::hup())).is_hup() {}

    Ok(().into())
}
```

Basically, if the socket matches any of the mask's readiness bits, then `Ready` is returned. If not, NotReady is returned and the task is scheduled to be notified.

This is also similar to how [`poll_ready`](https://docs.rs/tokio-core/0.1.12/tokio_core/reactor/struct.PollEvented.html#method.poll_ready) works in tokio-core today, but without conflating the read and the write halves.

**Comment by olix0r** [comment]
@carllerche at face value, Option 3 sounds the cleanest to me.
